### PR TITLE
Add the 'tizen-common-web' module

### PR DESCRIPTION
**Pull request by SejoongDeJang**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by SejoongDeJang** [comment]
```
1> tizen-common-web failing:
1> Error: /home/runner/work/DefinitelyTyped/DefinitelyTyped/types/tizen-common-web/index.d.ts:1:1
1> ERROR: 1:1  npm-naming  Declaration file must have a matching npm package.
1> To resolve this error, either:
1> 1. Change the name to match an npm package.
1> 2. Add a Definitely Typed header with the first line
1> 

1> 
1> // Type definitions for non-npm package tizen-common-web-browser
1> 
1> Add -browser to the end of your name to make sure it doesn't conflict with existing npm packages.
1> If you won't fix this error now or you think this error is wrong,
1> you can disable this check by adding the following options to your project's tslint.json file under "rules":
1> 
1>     "npm-naming": false
```
Refer to above build logs, my package is not released to the npm.
But my package [@tizentv/tizen-common-web](https://www.npmjs.com/package/@tizentv/tizen-common-web) is released to v1.0.0
Can you explain about it?

File: types/tizen-common-web/tslint.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    "extends": "dtslint/dt.json",
+    "rules":{
+        "no-const-enum": false,
```
Comment (weswigham) [inline types/tizen-common-web/tslint.json:4]: This rule is on for a very good reason - when you use a const enum, the const is directly compiled into your code, thus quietly burning a dependency on the version of the enum used at _build_ time into your application or library when used in an expression position (which is not guaranteed to match with the version used at _run_ time). For a final, built application, this is _OK_, but on DT, we can't say if consumers will be building a library or a complete application (that's up to them) - so we have to be very cautious. We offer string literal completions for the union members - can't those be good enough?

In short: Please don't disable this rule. Some of the others I can abide disabling (like line length), at least at first, for a large package like this, but this one creates a big usability issue for type consumers. (Unless those enum objects are actually available at runtime, in which case, normal non-const enums should be used, so the values are "dynamically" selected at runtime)

File: types/tizen-common-web/tslint.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+        "max-line-length": false,
+        "strict-export-declare-modifiers": false,
```
Comment (weswigham) [inline types/tizen-common-web/tslint.json:6]: It'd be nice if this could be cleaned up and this exception removed before merge. `declare interface` seems particularly egregious to me - there's no reason for an `interface` to have a `declare` modifier, ever.

File: types/tizen-common-web/lib/tvinfo.d.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    | 'CAPTION_EDGE_NONE'
+    | 'CAPTION_EDGE_RAISED'
+    | 'CAPTION_EDGE_DEPRESSED'
+    | 'CAPTION_EDGE_UNIFORM'
+    | 'CAPTION_EDGE_DROP_SHADOWED';
+/**
+ * All available values for the caption menu.
+ */
+declare namespace Tizen {
+    type CaptionValue = CaptionStateUnion &
```
Comment (weswigham) [inline types/tizen-common-web/lib/tvinfo.d.ts:260]: This is an intersection of unions of literal types, and is equivalent to `never`, if I'm not mistaken. I don't think that's the intent?

**Review by weswigham — CHANGES_REQUESTED** [review]

File: types/tizen-common-web/lib/tvinfo.d.ts (hunk 3 above)
Comment (SejoongDeJang) [reply types/tizen-common-web/lib/tvinfo.d.ts]: Before discuss about it, I modify the `CaptionStateUnion` to `CaptionUnion`. We want to support to developer about both a `string` type and a enumeration. But the `xxxxUnion` make it misunderstood, so we removed the `xxxxUnion`.

Now let's talk about the intersection.
```js
interface TVInfoManager {
        /**
         * Method returns the value for corresponding caption menu key.
         * @returns CaptionValue value for given caption menu key
         * @param key Caption menu
         * @throw WebAPIException TypeMismatchError, UnknownError
         * @since 2.4
         */
        getCaptionValue(key: CaptionInfoKeyUnion): CaptionValue;
```

Refer to the `getCaptionValue`, this method return a string value of the enum which is `CaptionStateUnion`, `    CaptionModeUnion`, `CaptionFontSizeUnion`, `CaptionFontStyleUnion`, `CaptionColorUnion`, `CaptionOpacityUnion` or `     CaptionEdgeUnion`. Thus, I think using intersection of enum is better than enumerate `string`s.

File: types/tizen-common-web/tslint.json (hunk 2 above)
Comment (SejoongDeJang) [reply types/tizen-common-web/tslint.json]: I understood. Please check the latest commit.

File: types/tizen-common-web/tslint.json (hunk 1 above)
Comment (SejoongDeJang) [reply types/tizen-common-web/tslint.json]: I understood. Please check the latest commit.

**Comment by SejoongDeJang** [comment]
@weswigham
I got a question about importing `interface` or `enum`.

Before discuss about it, I need a explain about our development environments. The [@tizentv/tizen-common-web](https://www.npmjs.com/package/@tizentv/tizen-common-web) is run only on Tizen OS, and it support device API such as filesystem, tvinfo and so on. Without the this package, the developer call the Tizen API with global `tizen` variable like a below

```js
//some.js
tizen.filesystem.openFile('documents/file', 'w');
```

It's totally outdated, but we can't refactoring it because of supporting legacy product.

So we made the [@tizentv/tizen-common-web](https://www.npmjs.com/package/@tizentv/tizen-common-web) to support commonjs style and increase the effidiency.

```js
//some.js
const { filesystem } = require('@tizentv/tizen-common-web);
filesystem.openFile("documents/file", "w");
```

In this modification, when I tested, I got error like

```
ERROR in ./spec/common/filesystem.spec.ts
Module not found: Error: Can't resolve 'filesystem' in 'D:\workspace\tsfy\ts-tizen-web\spec\common'
 @ ./spec/common/filesystem.spec.ts 2:0-103 71:19-27 72:19-27 73:19-27 74:19-27 75:19-27 78:19-40 79:19-40 82:19-41 83:19-41 84:19-41 87:19-35 88:19-35 89:19-35 93:63-71 98:69-77 103:76-84 245:46-54 253:15-23
 @ multi ./spec/common/application.spec.ts ./spec/common/download.spec.ts ./spec/common/filesystem.spec.ts ./spec/common/systeminfo.spec.ts ./spec/common/tizen.spec.ts ./spec/common/tvaudiocontrol.spec.ts ./spec/common/tvdisplaycontrol.spec.ts ./spec/common/tvinfo.spec.ts ./spec/common/tvinputdevice.spec.ts ./spec/common/tvwindow.spec.ts ./spec/common/websetting.spec.ts
```

Here is my test code.

```ts
import { filesystem } from '@tizentv/tizen-common-web';
import { FileMode } from 'filesystem';

// error!!!
filesystem.openFile('documents/file', FileMode.w);
```

Using `enum` or `interface` occur a webpack build error. I checked the `tsc --traceResolution` it was matached well. Do you why I got a webpack build error?

```
Resolving module name 'filesystem' relative to base url 'D:/workspace/tsfy/ts-tizen-web/' - 'D:/workspace/tsfy/ts-tizen-web/filesystem'.
File 'D:/workspace/tsfy/ts-tizen-web/filesystem.ts' does not exist.
File 'D:/workspace/tsfy/ts-tizen-web/filesystem.tsx' does not exist.
File 'D:/workspace/tsfy/ts-tizen-web/filesystem.d.ts' does not exist.
File 'D:/workspace/tsfy/types-tizen-common-web/filesystem.ts' does not exist.
File 'D:/workspace/tsfy/types-tizen-common-web/filesystem.tsx' does not exist.
File 'D:/workspace/tsfy/types-tizen-common-web/filesystem.d.ts' exist - use it as a name resolution result.
======== Module name 'filesystem' was successfully resolved to 'D:/workspace/tsfy/types-tizen-common-web/filesystem.d.ts'. ========
```

By the way, in this modification I read the `node` and our package is enable like a `node`. How to support multi module entry? Import the api from each modules. Is it possible?

```ts
import { FileMode, openFile } from 'filesystem';

openFile('documents/file', FileMode.w);
```

**Comment by weswigham** [comment]
> By the way, in this modification I read the node and our package is enable like a node. How to support multi module entry? Import the api from each modules.

Consumers need to include the package containing all the module definitions via the compiler's `types` option. Or add a `/// <refrerence types="containing-packagename" />` directive (like we do on DT).

File: types/tizen-common-web/lib/tvinfo.d.ts (hunk 3 above)
Comment (weswigham) [reply types/tizen-common-web/lib/tvinfo.d.ts]: You'd want a union of the enums, then, no? Since it can return a value from any one of the enums?

Comment (SejoongDeJang) [reply types/tizen-common-web/lib/tvinfo.d.ts]: The `getCaptionValue` return any one of the enums. 
And I got mistake, the `getCaptionValue` returns a intersection of enums like 
``` ts
type CaptionValue = CaptionState &
        CaptionMode &
        CaptionFontSize &
        CaptionFontStyle &
        CaptionColor &
        CaptionOpacity &
        CaptionEdge;
```

**Comment by sandersn** [comment]
@SejoongDeJang are you ready for this PR to be merged? I couldn't tell if @weswigham's comment addressed your latest problem with mismatch between compile-time and runtime imports.

**Comment by SejoongDeJang** [comment]
@sandersn I think, yes I'm ready.
